File: Kalorie/Kalorie/Core/UseCases/SearchFoodItemsUseCase.swift
//
//  SearchFoodItemsUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 22.07.2026.
//

import Foundation

protocol SearchFoodItemsUseCaseProtocol {
    func callAsFunction(query: String) async throws -> [FoodItemDomain]
}

struct SearchFoodItemsUseCase: SearchFoodItemsUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol) {
        self.dataProvider = dataProvider
    }

    // MARK: - Functions

    func callAsFunction(query: String) async throws -> [FoodItemDomain] {
        let lowercasedQuery = query.lowercased()
        async let byName: [FoodItemDTO] = dataProvider.loadAsync(
            from: Constants.Firestore.foodItems,
            where: "cz_name_lowercase",
            hasPrefix: lowercasedQuery,
            limit: 10
        )
        async let byOriginalName: [FoodItemDTO] = dataProvider.loadAsync(
            from: Constants.Firestore.foodItems,
            where: "eng_name_lowercase",
            hasPrefix: lowercasedQuery,
            limit: 10
        )
        let (nameResults, originalNameResults) = try await (byName, byOriginalName)
        var seen = Set<String>()
        return (nameResults + originalNameResults)
            .filter { seen.insert($0.id).inserted }
            .map { $0.asDomain() }
    }
}

#if DEBUG
struct SearchFoodItemsUseCaseFake: SearchFoodItemsUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false
    var stubbedItems: [FoodItemDomain] = []

    // MARK: - Functions

    func callAsFunction(query: String) async throws -> [FoodItemDomain] {
        if shouldThrow { throw NSError(domain: "SearchFoodItemsUseCaseFake", code: 0) }
        return stubbedItems
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/SearchFoodItemsUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemDTO
import antoni.kalorie.core.networking.loadArrayContainsAsync
import antoni.kalorie.core.networking.loadHasPrefixAsync
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.textkit.searchQuery
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

interface SearchFoodItemsUseCaseProtocol {
    suspend operator fun invoke(query: String): List<FoodItemDomain>
}

class SearchFoodItemsUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
) : SearchFoodItemsUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(query: String): List<FoodItemDomain> = coroutineScope {
        val searchQuery = searchQuery(query)
        val lowercasedQuery = searchQuery.lowercased
        val foldedQuery = searchQuery.folded
        val foldedLastWord = searchQuery.lastWord
        val byName = async { loadByPrefix("cz_name_lowercase", lowercasedQuery) }
        val byOriginalName = async { loadByPrefix("eng_name_lowercase", lowercasedQuery) }
        val byFoldedName = async { loadByPrefix("cz_name_folded", foldedQuery) }
        val byFoldedOriginalName = async { loadByPrefix("eng_name_folded", foldedQuery) }
        val byCzNameToken = async { loadByToken("cz_name_search_terms", foldedLastWord) }
        val byEngNameToken = async { loadByToken("eng_name_search_terms", foldedLastWord) }
        val seen = mutableSetOf<String>()
        listOf(byName, byOriginalName, byFoldedName, byFoldedOriginalName, byCzNameToken, byEngNameToken)
            .flatMap { it.await() }
            .filter { seen.add(it.id) }
            .map { it.asDomain() }
    }

    // MARK: - Private

    private suspend fun loadByPrefix(field: String, prefix: String): List<FoodItemDTO> =
        dataProvider.loadHasPrefixAsync(from = Constants.Firestore.FOOD_ITEMS, field = field, hasPrefix = prefix, limit = RESULT_LIMIT)

    private suspend fun loadByToken(field: String, token: String): List<FoodItemDTO> =
        dataProvider.loadArrayContainsAsync(from = Constants.Firestore.FOOD_ITEMS, field = field, arrayContains = token, limit = RESULT_LIMIT)

    private companion object {
        const val RESULT_LIMIT = 10
    }
}
